<a href="https://colab.research.google.com/github/abhinabasaha/Implementation_of_ANN_with_Keras/blob/master/ANN_with_Keras.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
#Mounting file to Google Co-laboratory

from google.colab import drive
drive.mount('/content/drive/')

Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount("/content/drive/", force_remount=True).


**Step 1: Data pre-processing**

In [0]:
# Importing the libraries

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd


In [0]:
#Importing the dataset

dataset = pd.read_csv("/content/drive/My Drive/DeepLearning/ANN_with_Keras/Artificial_Neural_Networks/Churn_Modelling.csv")
X = dataset.iloc[:, 3:13].values
y = dataset.iloc[:, 13].values

In [0]:
# Encoding categorical data (countries and gender)

from sklearn.preprocessing import LabelEncoder, OneHotEncoder

labelencoder_X_1 = LabelEncoder()
X[:, 1] = labelencoder_X_1.fit_transform(X[:, 1])



In [0]:
labelencoder_X_2 = LabelEncoder()
X[:, 2] = labelencoder_X_2.fit_transform(X[:, 2])


In [6]:
#Creating dummy variables for the 1st categorical data(corresponding to countries)

onehotencoder = OneHotEncoder(categorical_features = [1])
X = onehotencoder.fit_transform(X).toarray()

#Removing one coloumn to avoid dummy variable trap
X=X[:,1:]

/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/_encoders.py:415: FutureWarning: The handling of integer data will change in version 0.22. Currently, the categories are determined based on the range [0, max(values)], while in the future they will be determined based on the unique values.
If you want the future behaviour and silence this warning, you can specify "categories='auto'".
In case you used a LabelEncoder before this OneHotEncoder to convert the categories to integers, then you can now use the OneHotEncoder directly.
  warnings.warn(msg, FutureWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/_encoders.py:451: DeprecationWarning: The 'categorical_features' keyword is deprecated in version 0.20 and will be removed in 0.22. You can use the ColumnTransformer instead.
  "use the ColumnTransformer instead.", DeprecationWarning)


In [0]:
# Splitting the dataset into the Training set and Test set
    
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

In [0]:
# Feature Scaling

from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

 **Step 2: Creating the ANN**

In [9]:
#Importing the necessary libraries

import keras
from keras.models import Sequential
from keras.layers import Dense

Using TensorFlow backend.


In [10]:
# Initialising the ANN

classifier = Sequential()

W0711 08:51:13.412356 140284180449152 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:74: The name tf.get_default_graph is deprecated. Please use tf.compat.v1.get_default_graph instead.



In [11]:
# Adding the input layer and the first hidden layer

classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu', input_dim = 11))


W0711 08:51:13.446708 140284180449152 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:517: The name tf.placeholder is deprecated. Please use tf.compat.v1.placeholder instead.

W0711 08:51:13.455091 140284180449152 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:4138: The name tf.random_uniform is deprecated. Please use tf.random.uniform instead.



In [0]:

# Adding the second hidden layer

classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu'))


In [0]:
# Adding the output layer

classifier.add(Dense(units = 1, kernel_initializer = 'uniform', activation = 'sigmoid'))



In [14]:
# Compiling the ANN

classifier.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

W0711 08:51:13.539217 140284180449152 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/optimizers.py:790: The name tf.train.Optimizer is deprecated. Please use tf.compat.v1.train.Optimizer instead.

W0711 08:51:13.571686 140284180449152 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:3376: The name tf.log is deprecated. Please use tf.math.log instead.

W0711 08:51:13.580366 140284180449152 deprecation.py:323] From /usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/nn_impl.py:180: add_dispatch_support.<locals>.wrapper (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where


In [15]:
# Fitting the ANN to the Training set

classifier.fit(X_train, y_train, batch_size = 10, epochs = 100)

W0711 08:51:13.844362 140284180449152 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:986: The name tf.assign_add is deprecated. Please use tf.compat.v1.assign_add instead.



Epoch 1/100
8000/8000 [==============================] - 8s 946us/step - loss: 0.5042 - acc: 0.7956
Epoch 2/100
8000/8000 [==============================] - 4s 446us/step - loss: 0.4267 - acc: 0.7960
Epoch 3/100
8000/8000 [==============================] - 4s 445us/step - loss: 0.4189 - acc: 0.8166
Epoch 4/100
8000/8000 [==============================] - 4s 452us/step - loss: 0.4135 - acc: 0.8304
Epoch 5/100
8000/8000 [==============================] - 4s 446us/step - loss: 0.4094 - acc: 0.8327
Epoch 6/100
8000/8000 [==============================] - 3s 437us/step - loss: 0.4064 - acc: 0.8346
Epoch 7/100
8000/8000 [==============================] - 4s 452us/step - loss: 0.4047 - acc: 0.8339
Epoch 8/100
8000/8000 [==============================] - 4s 443us/step - loss: 0.4032 - acc: 0.8357
Epoch 9/100
8000/8000 [==============================] - 4s 443us/step - loss: 0.4023 - acc: 0.8351
Epoch 10/100
8000/8000 [==============================] - 4s 448us/step - loss: 0.4014 - acc: 0.8331

In [0]:
# Predicting the Test set results

y_pred = classifier.predict(X_test)
y_pred = (y_pred > 0.5)


In [22]:
y_pred

array([[False],
       [False],
       [False],
       ...,
       [False],
       [False],
       [False]])

In [0]:
# Making the Confusion Matrix

from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

In [23]:
cm

array([[1523,   72],
       [ 221,  184]])

Accuracy= (Number of Correct Predictions)/(Total number of predictions)

In [25]:
accuracy=(1523+184)/(2000)
accuracy*100

85.35000000000001

So we get an accuracy of 85.35% on new observations.

# Step 3:  Evaluating, Improving and Tuning the ANN

In [0]:
# Evaluating the ANN

from keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import cross_val_score
from keras.models import Sequential
from keras.layers import Dense
def build_classifier():
    classifier = Sequential()
    classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu', input_dim = 11))
    classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu'))
    classifier.add(Dense(units = 1, kernel_initializer = 'uniform', activation = 'sigmoid'))
    classifier.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])
    return classifier
classifier = KerasClassifier(build_fn = build_classifier, batch_size = 10, epochs = 100)
accuracies = cross_val_score(estimator = classifier, X = X_train, y = y_train, cv = 10, n_jobs = -1)
mean = accuracies.mean()
variance = accuracies.std()


In [21]:
# Improving the ANN
# Dropout Regularization to reduce overfitting if needed

# Tuning the ANN
from keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import GridSearchCV
from keras.models import Sequential
from keras.layers import Dense
def build_classifier(optimizer):
    classifier = Sequential()
    classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu', input_dim = 11))
    classifier.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu'))
    classifier.add(Dense(units = 1, kernel_initializer = 'uniform', activation = 'sigmoid'))
    classifier.compile(optimizer = optimizer, loss = 'binary_crossentropy', metrics = ['accuracy'])
    return classifier
classifier = KerasClassifier(build_fn = build_classifier)
parameters = {'batch_size': [25, 32],
              'epochs': [100, 500],
              'optimizer': ['adam', 'rmsprop']}
grid_search = GridSearchCV(estimator = classifier,
                           param_grid = parameters,
                           scoring = 'accuracy',
                           cv = 10)
grid_search = grid_search.fit(X_train, y_train)
best_parameters = grid_search.best_params_
best_accuracy = grid_search.best_score_

Epoch 1/100
7200/7200 [==============================] - 2s 212us/step - loss: 0.5577 - acc: 0.7960
Epoch 2/100
7200/7200 [==============================] - 1s 177us/step - loss: 0.4350 - acc: 0.7971
Epoch 3/100
7200/7200 [==============================] - 1s 176us/step - loss: 0.4291 - acc: 0.7971
Epoch 4/100
7200/7200 [==============================] - 1s 175us/step - loss: 0.4260 - acc: 0.7971
Epoch 5/100
7200/7200 [==============================] - 1s 179us/step - loss: 0.4235 - acc: 0.7971
Epoch 6/100
7200/7200 [==============================] - 1s 177us/step - loss: 0.4213 - acc: 0.7971
Epoch 7/100
7200/7200 [==============================] - 1s 181us/step - loss: 0.4193 - acc: 0.8062
Epoch 8/100
7200/7200 [==============================] - 1s 174us/step - loss: 0.4175 - acc: 0.8211
Epoch 9/100
7200/7200 [==============================] - 1s 185us/step - loss: 0.4163 - acc: 0.8235
Epoch 10/100
7200/7200 [==============================] - 1s 177us/step - loss: 0.4155 - acc: 0.8264

KeyboardInterrupt: ignored